### DISCUSSION TITLE
How do I implement exportToCookie on pocketbase server?

### DISCUSSION THREAD

**Discussion by Knorway** [pocketbase/pocketbase#3051] [Q&A] [answered]
I using sveltekit and following the instructions here(https://github.com/pocketbase/js-sdk#ssr-integration).
It would be much easier if I register user on server but since I’m using oauth2 on client side(don’t want to exchange code manually 😂), I’m doing the followings to inject oauth2 result to cookie initially:

```
<script lang="ts">
	import Pocketbase from 'pocketbase';

	const googleLogin = async () => {
		const pb = new Pocketbase('http://127.0.0.1:8090');
		try {
			await pb.collection('users').authWithOAuth2({ provider: 'google' });
			await fetch('/api/cookie', {
				method: 'POST',
				credentials: 'include',
				body: pb.authStore.exportToCookie()
			});
			location.href = '/';
		} catch (error) {
			pb.authStore.clear();
			console.log(error);
		}
	};
</script>

<button on:click={googleLogin}>login with google</button>
```
```
export const POST = async ({ request }) => {
	return new Response(null, {
		status: 200,
		headers: {
			'set-cookie': await request.text()
		}
	});
};
```
This workaround works but since I’m using pocketbase as a framework, is it possible to send cookie to user in the first place on pocketbase server? So the question is, can I make the same cookie string with client’s exportToCookie method on server?(pb_auth={token: ….}; Path=‘/‘, ….)
```
	pb.App.OnRecordAfterAuthWithOAuth2Request().Add(func(e *core.RecordAuthWithOAuth2Event) error {
		cookie := http.Cookie{
			// can I do something here?
		}
		e.HttpContext.SetCookie(&cookie)
		return nil
	})
```
thank you for you time and the awesome project!

**Comment by ganigeorgiev** [comment]
PocketBase doesn't use cookies out of the box. The authentication happens only if the request has `Authorizaiton: TOKEN` header.

Note that even if you write the cookie in the OAuth2 success response, you still will have to load it in the SDK `AuthStore` so that its token can be included later with the `Authorization` header.

Instead of creating a cookie loader middleware in Go, and since you already seems to have mixed PocketBase server access (aka. in the browser and node), you can try to export the cookie with `exportToCookie({ httpOnly: false })`. This practically will allow access to the cookie from both server and client-side JS, which means that in your client-side SDK instance you can attach to the `authStore.onChange` listener and leave the browser to set the cookie with `document.cookie` or other cookie abstraction. Something like:

```js
// client side SDK instance

import PocketBase from "pocketbase";

const pb = new PocketBase("https://example.com");

pb.authStore.onChange((token, model) => {
    document.cookie = pb.authStore.exportToCookie({ httpOnly: true });
});

...

await pb.collection("users").authWithOAuth2({ ... })
```

**Comment by Knorway** [comment]
thank you for the comment gani. yes, I'm doing this only to load from it in order to use in sveltekit server.
```
// src/hooks.server.ts
export async function handle({ event, resolve }) {
    event.locals.pb = new PocketBase("http://127.0.0.1:8090");
    event.locals.pb.authStore.loadFromCookie(event.request.headers.get('cookie') || ''); // <- everything is for this to work
    ...
}
```
and I know that pb has got nothing to with understanding cookie or making one. 
I was just wondering if I can make equivalent cookie string in golang server if there are supported admin api, but it seems not.
and to be honest, i will probably not use httpOnly: false at any cost😂
thank you for you comment once again! pb is awesome!

**Reply by ganigeorgiev** [reply]
> I was just wondering if I can make equivalent cookie string in golang server if there are supported admin api, but it seems not.

There is no builtin Go apis for generating the same auth cookie as the JS SDK, but the generated cookie string is nothing special - it is serialized JSON object with 2 keys `token` and `model` _(optional)_.

> will probably not use httpOnly: false at any cost

If security is your concern - it should be safe to set `httpOnly: false` as long as:
- you have a basic [CSP configured](https://github.com/pocketbase/js-sdk#security)
- you don't load untrusted 3rd party JS

**Reply by Knorway** [reply]
thank you for the instructions. I think I'm getting closer. I found tokens.NewRecordAuthToken() returns same token unless user record's token key is unchanged. to save your time, followings will be the last questions for this one:

1. using tokens.NewRecordAuthToken() for this purpose looks okay to you?
2. can record.RefreshTokenKey() be used to invalid already generated tokens? will pb reject old tokens after reset the token key? I remember once you mentioned that changing password is only way to invalid pre generated tokens.

**Reply by ganigeorgiev** [reply, marked as answer]
For 1) - Yes, it is fine to use it.

For 2 - Yes, `record.RefreshTokenKey()` will invalid all previously issued tokens for the specified record model but note that you need also to persist the model change in the database, aka. you'll have to call also `app.Dao().SaveRecord(record)`

> I remember once you mentioned that changing password is only way to invalid pre generated tokens.

Changing the password (aka. `record.SetPassword()`) calls `record.RefreshTokenKey()` under the hood.

**Reply by Knorway** [reply]
thank you so much gani!
